pkg/agent/flowexporter/exporter/exporter.go

```diff
@@ -295,19 +291,14 @@ func (exp *flowExporter) sendFlowRecords() error {
 				}
 				exp.numDataSetsSent = exp.numDataSetsSent + 1
 			}
-			// If the connection is not present in the conntrack table anymore,
-			// then we delete the connection in connection store and record store.
-			// Please note that our record cache policy is that we keep the record in
-			// the record map as long as the corresponding connection is present in
-			// the conntrack table.
-			if !record.Conn.IsPresent {
-				klog.V(2).Infof("Deleting the inactive connection with key: %v from both connection map and record map", key)
-				if err := exp.connStore.DeleteConnectionByKey(key); err != nil {
-					return err
-				}
+			// If connection is in dying state, we will delete all the flow records.
+			if flowexporter.IsConnectionDying(&record.Conn) {
+				klog.V(2).Infof("Deleting the inactive connection with key: %v from record map", key)
 				if err := exp.flowRecords.DeleteFlowRecordWithoutLock(key); err != nil {
 					return err
 				}
+				conn, _ := exp.connStore.GetConnByKey(key)
+				conn.IsExportedAndDying = true
```

Comment:
Will this be actually set in the connection map of ConnectionStore? Don't we have to assign the connection again?
We may have to modify the method `addOrUpdateConn(conn *flowexporter.Connection)` and make this public.

Could we add a unit test for the scenario where `conn.IsExportedAndDying` flag is set and we do not do extra processing on it?

Reply:
I went through the code again and considered we can simplify the optimization as follows:

1. handle connection deletion from connection map in resetConn from `Poll()` when the connection no longer exists in conntrack table (`connections.go`)
2. delete flow records of corresponding dying connection after exporting (`exporter.go`)
3. avoid updating connection map and flow records map when the connection is in dying state
In this scenario, we don’t need to have separate variable like `IsTerminatedAndExported` because when connection comes in as dying state, it will always be added to connection map, and it will only be removed when conntrack table does not contain this connection. And the flow records of dying connections will always be deleted after exported.

Reply:
Just to confirm. Here "dying state" corresponds to the new utility you have right, i.e., for TCP protocol monitor TIME_WAIT and CLOSE and for other protocols DYING status flag?

Reply:
Correct.

Reply:
Ok. The optimization logic sounds good to me. Please update the PR description briefly explaining what optimization is being done here.

Reply:
Done. Thanks!